File: coding_everyday/lc53/Maximum Subarray.py
import bisect
import math
from typing import List
from collections import defaultdict


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        dp = nums[0]
        maax = dp
        for i in range(1, n):
            dp = max(nums[i] + dp, nums[i])
            maax = max(maax, dp)
        return maax


if __name__ == "__main__":
    sol = Solution()
    nums = [5,4,-1,7,8]
    nums = [-1, -2]
    nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
    print(sol.maxSubArray(nums))

File: coding_everyday/lc1-100/lc53/src/MaximumSubarray.java
public class MaximumSubarray {
    public static void main(String[] args) {
//        int[] nums = {-2,1,-3,4,-1,2,1,-5,4};
//        int[] nums = {1};
//        int[] nums = {5,4,-1,7,8};
        int[] nums = {1, 2};
        Solution sol = new Solution();
        System.out.println(sol.maxSubArray(nums));
    }
}

class Solution {
    public int maxSubArray(int[] nums) {
        int sum = nums[0];
        int max = nums[0];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] > nums[i] + sum) {
                sum = nums[i];
            } else {
                sum += nums[i];
            }
            max = Math.max(sum, max);
        }
        return max;
    }
}